### ISSUE TITLE
Support:  queue  there is empty/drain and there is idle vs ?

### ISSUE THREAD

**Issue by ORESoftware** [caolan/async#1547]
Ok so when `empty()` fires it means the last item _started_ processing. When `drain()` fires, the last item finished processing.

So, I assume `idle()` returns true when the last item finished processing or when the last item started processing?

Is there another call besides idle that we can use? **What I am looking for is** a check to see if the queue has finished processing all items, or if the queue is completely empty.

On the same note - I am looking for a count of the number of items that have been pushed/processed by the queue.

Basically, the **problem I am ultimately trying to solve** is how to handle the case where the queue was never passed any items to process. In that case, drain() / error() will never happen, so I need to be able to fire another callback saying there were no items to process.

https://github.com/ORESoftware/node-check-fast/blob/master/src/ncf.ts

**Comment by hargasinski** [comment]
> So, I assume idle() returns true when the last item finished processing or when the last item started processing?

The former is true. `q.idle()` will return `true` if there are no items being processed and no items waiting to be processed.

> Is there another call besides idle that we can use? What I am looking for is a check to see if the queue has finished processing all items, or if the queue is completely empty.

`q.idle()` will work for this use case. It will return `true` if the queue is completely empty, i.e. it finished processing all of its items, or no items have been pushed.

```js
let q = async.queue(function(task, callback) {
    // do some work
    callback()
}, 1)

// true
q.idle()
```

> On the same note - I am looking for a count of the number of items that have been pushed/processed by the queue.

We don't keep track of the number of items processed by the queue. It's relatively easy to add yourself. Something along the lines of:

```js
let numTasksStarted = 0
let q = async.queue(function(task, callback) {
  ++numTasksStarted
  // do some work
  callback()
}, 1)

// numTasksStarted is the total number of tasks that have started processing
// numTasksStarted - q.running() is the total number of tasks that have finished processing
// numTasksStarted + q.length() is is the total number of tasks processed/being processed/waiting to be processed
```

> Basically, the problem I am ultimately trying to solve is how to handle the case where the queue was never passed any items to process. In that case, drain() / error() will never happen, so I need to be able to fire another callback saying there were no items to process.

When you are done waiting for data, for example, in your case, it might be on the `end` or `exit` event, you can check `q.started`. If `q.started` is `false`, no items were pushed to the queue.

See the [`QueueObject` docs](https://caolan.github.io/async/docs.html#QueueObject) for more info.

**Comment by ORESoftware** [comment]
@hargasinski  thanks, let me digest this, will probably close it tomorrow if I don't have any more questions
